### PR TITLE
Adding LaunchDarkly example

### PR DESCRIPTION
**Pull request by remotesynth**
This adds a LaunchDarkly example under the typescript section. Please let me know if everything is in the format expected.

### CODE REVIEW

**Comment by brnkrygs** [comment]
Hi @remotesynth! 

Thanks for your contribution. We'll get it into our review pipeline and will comment on this PR if there are any changes to discuss.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json:23]: Incorrect file path (I think?)

File: typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json:14]: Can we call these Lambda _functions_?  I.E. "Lambda function using LaunchDarkly"

File: typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json:18]: "LaunchDarkl" missing 'y'

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:1]: Can we rename to README (all caps) for consistency?

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:5]: Please call them Lambda _functions_...

>... integrating feature flags into your Lambda _functions_ using the ...

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:21]: Incorrect heading... or maybe not? This section seems to belong after "About this Example", maybe as a subheading in there? It seems to assume I understand the "About this Example" section.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:25]: Consider adding links from the function name to the code.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:35]: canbe --> can be

Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:35]: Great paragraph, I appreciate how you describe the ways feature flags can let you test code paths in various environments... *without* enabling them for all users.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:38]: Can we be explicit in mapping to testing on this bullet? Something like:

>Only particular users or segments of users, such as a test identity

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:72]: Can you add a note about having an authenticated CLI session as well?

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:87]: Lets keep this file path consistent with the rest - the other examples just use `cd cdk` and `cd create-layer`

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:45]: Good intro to the SDKs!

Can we be explicit about a LaunchDarkly account being needed (or not needed) to run the example? It may also be a good idea to add a link to LaunchDarkly "Getting Started" content if customers want to take the next step.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:59]: Please link "Lambda Layer" to the AWS docs.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:63]: Thanks for including this script. Looks like there's a `layer.zip` included in the repo under `resources/layers`. Can we drop that since users will be creating the layer itself (thinking of managing overall repo size here).

File: typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/getFlagsStarter/index.ts
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/getFlagsStarter/index.ts:8]: This should be set by the CDK deployment. Please add a placeholder there that users can update before they deploy (looks like all functions need it). Please add instructions in the README to update it.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/getFlagsStarter/index.ts
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/getFlagsStarter/index.ts:37]: Lets be explicit about this new code being the code path that you want to test before a full-scale rollout.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/getFlagsWithDynamo/index.ts
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/getFlagsWithDynamo/index.ts:11]: Please add environment variable in the CDK template (unless I'm missing it there?)

File: typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/getFlagsWithDynamo/index.ts
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/getFlagsWithDynamo/index.ts:23]: See other comment about adding env var to CDK.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/getFlagsWithDynamo/index.ts
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/getFlagsWithDynamo/index.ts:48]: See other comment about this being the code path under test.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/syncFlagsToDynamo/index.ts
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/syncFlagsToDynamo/index.ts:25]: Please see earlier comment about environment variables in CDK definition.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/syncFlagsToDynamo/index.ts
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/lambdas/syncFlagsToDynamo/index.ts:1]: Consider adding language about this being a backend maintenance process, and not one that should be included in the request path of (for example) a web endpoint handler.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/test/cdk.test.js
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/test/cdk.test.js:1]: This looks like it can be dropped.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/layers/layer.zip
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/cdk/resources/layers/layer.zip:1]: As mentioned on the readme file, lets drop this to manage the overall repo size. Users can create via the `create-layer` script.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/img/LaunchDarkly-AWS-diagram.png
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/img/LaunchDarkly-AWS-diagram.png:1]: It doesn't look like this solution has Amplify, Cognito, or API Gateway in it. Can we update the diagram to reflect the solution or the solution to reflect the diagram?

If not, we should be explicit about the disconnect.

**Review by brnkrygs — CHANGES_REQUESTED** [review]
Hi @remotesynth !

Thanks so much for your work here. Well-commented code and a solid walkthrough of getting up and running.

There's a few changes requested across the README and Lambda functions - mostly wording.

One thing I'd love to see though is an integration test suite that actually exercises one of the Lambda functions. That way readers can see how testing and LaunchDarkly can work well together.

For example, a suite of two tests that specify different users in request headers. Then the LD flag can trigger one path for one test and a different path for the other test.

Please let me know if you have any questions on the comments! Happy to consider other ideas or feedback!

File: typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json
Comment (remotesynth) [reply typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json:23]: They appear to be correct to me. The `.ts` files are also in there so that they can regenerate the files if they need to from the TypeScript source but you can't deploy those directly without compiling them.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json
Comment (remotesynth) [reply typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json]: Pushed a fix

File: typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json
Comment (brnkrygs) [reply typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json:23]: Just the third one... isn't that supposed to be the `syncFlagsToDynamo` function rather than `getFlagsStarter`?

Comment (remotesynth) [reply typescript-test-samples/launchdarkly-lambda-dynamodb/metadata.json]: Ah, yes. Sorry about that.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (remotesynth) [reply typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md]: Done

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (remotesynth) [reply typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:27]: > Only particular users or segments of users, for example a test identity or a segment containing multiple identities performing Q&A

File: typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md
Comment (remotesynth) [reply typescript-test-samples/launchdarkly-lambda-dynamodb/readme.md:1]: This is very strange. On my local repo it was all caps. GitHub never saw the change and I had to rename it and rename it back to force it, but seems to be ok now.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/img/LaunchDarkly-AWS-diagram.png
Comment (remotesynth) [reply typescript-test-samples/launchdarkly-lambda-dynamodb/img/LaunchDarkly-AWS-diagram.png:1]: I like the overview as it explains how this all works in a complete application but I have updated the illustration to make it clearer what pieces are in the example.

**Comment by remotesynth** [comment]
Thanks for all the feedback. I have addressed each of the individual items.

Regarding the test event, this would require a larger change. The current code does not look for identity information passed via JWT or in the event. As noted in the comments, it uses an anonymous user context, which could still be useful for experimentation or gradual rollouts.

To create tests that would target users by their identity would probably require an additional example with additional targeting code as well as test code that could further delay completing this example.

**Comment by brnkrygs** [comment]
> Thanks for all the feedback. I have addressed each of the individual items.
> 
> Regarding the test event, this would require a larger change. The current code does not look for identity information passed via JWT or in the event. As noted in the comments, it uses an anonymous user context, which could still be useful for experimentation or gradual rollouts.
> 
> To create tests that would target users by their identity would probably require an additional example with additional targeting code as well as test code that could further delay completing this example.

Thanks @remotesynth ! I'll take another review pass.

Regarding adding tests, would you be comfortable with someone from our serverless testing initiative taking a 2nd pass (after we merge this) to add them?

**Comment by brnkrygs** [comment]
> > Regarding adding tests, would you be comfortable with someone from our serverless testing initiative taking a 2nd pass (after we merge this) to add them?
> 
> Sure. I am happy to work with them on it and even come up with an additional example. I think it could be helpful.

Ok, great! I don't necessarily want to block this PR on it, but I do think it would be a really valuable addition.

File: typescript-test-samples/launchdarkly-lambda-dynamodb/README.md
Comment (brnkrygs) [inline typescript-test-samples/launchdarkly-lambda-dynamodb/README.md:72]: Small update... users need to run `npm install` in this directory before running `npm run build`.

Otherwise they get:

```
node:internal/modules/cjs/loader:1152
  throw err;
  ^

Error: Cannot find module 'archiver'
Require stack:
- /Users/.../serverless-test-samples/typescript-test-samples/launchdarkly-lambda-dynamodb/create-layer/index.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1149:15)
    at Module._load (node:internal/modules/cjs/loader:990:27)
    at Module.require (node:internal/modules/cjs/loader:1237:19)
    at require (node:internal/modules/helpers:176:18)
    at Object.<anonymous> (/Users/.../serverless-test-samples/typescript-test-samples/launchdarkly-lambda-dynamodb/create-layer/index.js:3:18)
    at Module._compile (node:internal/modules/cjs/loader:1378:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1437:10)
    at Module.load (node:internal/modules/cjs/loader:1212:32)
    at Module._load (node:internal/modules/cjs/loader:1028:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:142:12) {
  code: 'MODULE_NOT_FOUND',
```

Comment (brnkrygs) [reply typescript-test-samples/launchdarkly-lambda-dynamodb/README.md:72]: Also, not getting `cdk/resources/layers/layer.zip` created when I run `npm run build`. Not sure what's going on - maybe an async issue? Seems like the script runs and ends early without waiting for the archive file to be created.

Comment (brnkrygs) [reply typescript-test-samples/launchdarkly-lambda-dynamodb/README.md:72]: Looks like its failing because `cdk/resources/layers` doesn't exist.

Please either create the directory and use a `.gitkeep` file to check it in, or have the build script create it if it doesn't already exist.

I created the directory manually, and the zip file was created as expected.

**Review by brnkrygs — CHANGES_REQUESTED** [review]
Hi @remotesynth !

Got a chance to test out the updates. Thanks for all the changes.

There's two small tweaks we need to make yet (specified as comments on the README). They're blocking users from building/deploying the solution.

After that, I think we'll be good to merge!

Appreciate the effort here.

**Comment by remotesynth** [comment]
@brnkrygs I made the readme change and added the `.gitkeep` file in. I believe that resolves all of the open issues. Thanks again!

**Review by brnkrygs — APPROVED** [review]
LGTM! 🚀

**Pull request merged**
